src/engine/decoders/decoder.h

```
@@ -0,0 +1,48 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "../scheduled_requests.h"
+#include "../../models/model.h"
+#include "../cache_manager.h"
+
+namespace Generators {
+
+struct ModelIO : State {
+  ModelIO(const GeneratorParams& params, const Model& model) : State(params, model) {}
+
+  DeviceSpan<float> Run(int total_length, DeviceSpan<int32_t>& next_tokens, DeviceSpan<int32_t> next_indices = {}) override {
+    return {};
+  }
+
+  void RewindTo(size_t index) override { (void)index; };
```

Perhaps these functions should return an error rather than be non-op. In the future we could abstract out only the necessary parts of the State struct and inherit from that without these extra functions.